### ISSUE TITLE
Proposal: Mismatching parameter name lint

### ISSUE THREAD

**Issue by rickvanprim** [rust-lang/rust#74769]
This lint detects when the implementation of a trait names a function parameter differently from the name in the source declaration.  Exceptions to this are prefixing the name with `_` or replacing the name with `_` to suppress the `unused-variable` lint. Some example code that triggers this lint:

```
trait Foo {
    fn foo(my_foo: &Foo);
}

struct Bar;
impl Foo for Bar {
    fn foo(your_foo: &Foo) {
        // ...
    }
}
```

This will produce:

```
warning: the parameter `your_bar` has a different name than the trait definition
  --> src/main.rs:7:16
   |
7  |         fn foo(your_foo: &Foo) {,
   |                ^^^^^^^^
   |                |
   |                help: consider renaming this parameter to `my_foo`
   |
```

**Comment by Lonami** [comment]
Wouldn't this make the names somewhat "part of the API"? Say a library offers a trait that I use but then suddenly they change the name. Now my project is building with warnings were there were none before. I guess in some cases it can actually be desired though.

**Comment by rickvanprim** [comment]
It's something I want for the ability to detect the above case.  I think the lint would have to be allow by default for this reason, so it doesn't spam people with warnings unless this is something they care about.

**Comment by estebank** [comment]
This seems like it would have to be somehow opt in, in the same way that `#[must_use]` is opt in.

**Comment by rickvanprim** [comment]
@estebank is that not what allowed by default lints are?
https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html

**Comment by estebank** [comment]
@rickvanprim not exactly. I'm staying that there should be a way to annotate the traits fns opting into this lint. An allowed by default lint would otherwise be triggered by _every_ impl once enabled.

**Comment by rickvanprim** [comment]
My only concern with needing to annotate is that is for my codebase that now means that every single trait I author now needs to add an attribute above it.  Though if opting in could be done at a crate level, that wouldn't be a big deal.
